libs/shmpipe/shmpipe.c

```diff
@@ -0,0 +1,659 @@
+{
+	pthread_mutex_destroy(&event->mutex);
+	pthread_cond_destroy(&event->cond);
+}
+
+static void event_reset(event_t* event)
+{
+	event->signaled = 0;
+}
+
+static void event_signal(event_t* event)
+{
+	pthread_mutex_lock(&event->mutex);
+	event->signaled = 1;
+	pthread_cond_broadcast(&event->cond);
+	pthread_mutex_unlock(&event->mutex);
+}
+
+
+static void event_wait(event_t* event)
+{
+	pthread_mutex_lock(&event->mutex);
+	while (!event->signaled)
+		pthread_cond_wait(&event->cond, &event->mutex);
+	pthread_mutex_unlock(&event->mutex);
+}
+
+static void latch_acquire(latch_t* latch)
+{
+    latch_t unset = 0;
+ 	while (!atomic_compare_exchange_strong(latch, &unset, 1)) {
+		RELEASE_CPU();
+		unset = 0;
+	}
+}
+
+static void latch_release(latch_t* latch)
+{
+	atomic_store(latch, 0);
+}
+
+#define ALIGN(x,y) (((x) + (y) - 1) & ~((y) - 1))
+
+void shmem_pipe_process_request(pipe_t* pipe, char const* req, size_t req_size, char* resp)
+{
+	message_header_t req_hdr;
+	message_header_t resp_hdr;
+	int header_sent = 0;
+	int header_received = 0;
+	size_t resp_size = sizeof(resp_hdr);
+
+	latch_acquire(&pipe->req.cs);
+
+	/* append data to queue head */
+	while (req_size != 0)
+	{
+		/* To distinguish empty buffer from full buffer we need to reserve some space */
+		size_t available = (pipe->req.head.pos >= pipe->req.tail.pos
+					 ? QUEUE_BUF_SIZE - (pipe->req.head.pos - pipe->req.tail.pos)
+					 : pipe->req.tail.pos - pipe->req.head.pos) - MESSAGE_DATA_ALIGNMENT;
+		if (available == 0)
+		{
+			/* Ring buffer is full: wait until consumer takes some requests */
+#ifndef BUSY_WAIT_RESPONSES
+			if (header_sent && pipe->req.head.n_blocked != 0)
+			{
+				pipe->req.head.n_blocked = 0;
+				event_signal(&pipe->req.head.event); /* notify receiver that head is advanced */
+			}
+			do
+			{
+				/* Wait until tail is advanced */
+				pipe->req.tail.n_blocked += 1;
+				event_reset(&pipe->req.tail.event);
+				latch_release(&pipe->req.cs);
+				event_wait(&pipe->req.tail.event);
+				latch_acquire(&pipe->req.cs);
+			} while (!header_sent && pipe->req.busy);
+#else
+			do
+			{
+				/* Wait until tail is advanced */
+				latch_release(&pipe->req.cs);
+				RELEASE_CPU();
+				latch_acquire(&pipe->req.cs);
+			} while (!header_sent && pipe->req.busy);
+#endif
+		}
+		else
+		{
+			size_t tail = QUEUE_BUF_SIZE - pipe->req.head.pos;
+			assert(available >= MESSAGE_DATA_ALIGNMENT);
+			if (!header_sent)
+			{
+				assert(!pipe->req.busy);
+				req_hdr.id = ++pipe->msg_id; /* protected by req.mutex */
+				req_hdr.size = req_size;
+				assert(tail >= MESSAGE_DATA_ALIGNMENT);
+				memcpy(&pipe->req.data[pipe->req.head.pos], &req_hdr, sizeof req_hdr);
+				pipe->req.head.pos = (pipe->req.head.pos + sizeof req_hdr) % QUEUE_BUF_SIZE;
+				pipe->req.busy = 1; /* prevent interleving with other messages */
+				header_sent = 1;
+			}
+			else
+			{
+				if (available > req_size)
+					available = req_size;
+
+				if (tail >= available)
```

Comment:
I found this pretty confusing, even though it's correct.

The `tail` variable denotes is the **size** of the byte range between head index and buffer end.
But when I read `tail`, I expect it to denote a **position**.

I tried to find a better name for this variable, but I can't find one.
Because it kind of means different things depending on whether `tail >= available` or not.

Please find a better name for the variable.